clvm/run_program.py

```diff
@@ -54,8 +56,12 @@ def run_program(
         pre_eval_op = None
 
     def traverse_path(sexp: SExp, env: SExp) -> Tuple[int, SExp]:
-        cost = PATH_LOOKUP_COST_PER_LEG
-        if sexp.nullp():
+        cost = TRAVERSE_BASE_COST
+        atom = sexp.atom
+        while len(atom) and atom[0] == b"\0":
+            cost += TRAVERSE_COST_PER_BYTE
+            atom = atom[1:]
+        if len(atom) == 0:
```

Comment:
shouldn't this be a failure? The most significant bit, iirc, terminates the traversal. If you don't have a most significant bit, it seems reasonable to treat that as an error.

Reply:
Intuitively, I think of a string of 0 bytes as a non-minimal 0, which has a specific meaning (evaluates to nil). I'm not sure what the benefit of failure is. Maybe we have a strict mode for debugging and non-strict mode for consensus?

Reply:
> Intuitively, I think of a string of 0 bytes as a non-minimal 0, which has a specific meaning (evaluates to nil).

0 is not a valid path though, right?

I would think the benefit of failure are:

* it's a grey zone that's not necessary and likely to not be intuitive. Sort of like UTF-8 overlong representations being illegal.
* it's easier to make it well-formed if we come up with a reasonable meaning in the future, since no valid program will rely on it. I suppose we can't rely on all nodes accepting a new behavior though.

Reply:
`0` actually is valid... it used to be the same as `1`, but within the past few weeks, it's changed to evaluate to `0`, as an analog to the 0 register in some RISC CPUs. The compiler does not yet use this optimization.

Reply:
Also, I'm pretty sure the compiler currently uses non-minimal paths, which isn't really ideal.